File: HW/HW7/hw7.java
    public int minMeetingRooms(int[][] intervals) {
        if (intervals == null || intervals.length == 0) {
            return 0;
        }

        Arrays.sort(intervals, (a, b) -> a[0] - b[0]);

        PriorityQueue<Integer> endTimeMinHeap = new PriorityQueue<>();

        for (int[] interval : intervals) {
            if (!endTimeMinHeap.isEmpty() && endTimeMinHeap.peek() <= interval[0]) {
                endTimeMinHeap.poll();
            }
            endTimeMinHeap.offer(interval[1]);
        }

        return endTimeMinHeap.size();
    }

File: HW/HW7/hw7.py
def minMeetingRooms(self, intervals: List[List[int]]) -> int:
    if not intervals:
        return 0

    intervals.sort(key=lambda x: (x[0], x[1])) 

    endTimeMinHeap = []

    for interval in intervals:
        if endTimeMinHeap and endTimeMinHeap[0] <= interval[0]:
            heapq.heappop(endTimeMinHeap)
        heapq.heappush(endTimeMinHeap, interval[1])

    return len(endTimeMinHeap)
